Improve README.md (#5543)
realm/SwiftLint: README.md

![](https://raw.githubusercontent.com/realm/SwiftLint/main/assets/screenshot.png)

This project adheres to the [Contributor Covenant Code of Conduct](https://realm.io/conduct).
This project adheres to the
[Contributor Covenant Code of Conduct](https://realm.io/conduct).
By participating, you are expected to uphold this code. Please report
unacceptable behavior to [info@realm.io](mailto:info@realm.io).

> Language Switch: [中文](https://github.com/realm/SwiftLint/blob/main/README_CN.md), [한국어](https://github.com/realm/SwiftLint/blob/main/README_KR.md).
> Switch Language:
> [中文](https://github.com/realm/SwiftLint/blob/main/README_CN.md)
> [한국어](https://github.com/realm/SwiftLint/blob/main/README_KR.md)

## Video Introduction

To get a high-level overview of SwiftLint, we encourage you to watch this
presentation recorded January 9th, 2017 by JP Simard (transcript provided):

[![Presentation](https://raw.githubusercontent.com/realm/SwiftLint/main/assets/presentation.svg)](https://youtu.be/9Z1nTMTejqU)

## Installation

> ### Using [Swift Package Manager](https://github.com/apple/swift-package-manager):

Reviewer: I prefer the clickable name over the explicit URL.

Reviewer 2: This was something that Garric and I discussed a bit before proposing this new style. And I have just gone back and looked at the new rendered markdown again for a fresh perspective. I do still believe this new style is preferred for the following reasons:

- The word "Using" does not fit with all of the headings, so there would be a lack of consistency when other words are used on some of them.
- Not all have URLs, so again from a consistency perspective, some of the headings would have different colors than others, due to being clickable.
- Seeing the URLs provides readable metadata that is helpful when looking at each of them.

Reviewer: The thing is that bare URL are discouraged by style guides and (Markdown) linters and I also consider them as a distraction while reading. Differently colored headlines are okay I think. These links are anyway not that important but rather friendly pointers.

Reviewer 2: Oh, thanks for that clarification 👍

I like the idea of removing "Using" so the entire headline can be clickable.

SwiftLint can be used as a [command plugin](#swift-package-command-plugin)
or a [build tool plugin](#swift-package-build-tool-plugins).

Add the following to your `Package.swift` file:

```swift
.package(url: "https://github.com/realm/SwiftLint.git", from: "<version>")
```

Or to pin to an exact version:

```swift
.package(url: "https://github.com/realm/SwiftLint.git", exact: "<version>")
```

> [!IMPORTANT]
> Replace `<version>` with the desired minimum or exact version.

> ### Xcode Package Dependency

Reviewer: Isn't this only useful to develop something against SwiftLint, like an own version with own rules? That's very advanced so probably shouldn't be mentioned here.

Reviewer 2: This is **_not_** for developing something against SwiftLint.

The existence of SwiftLint as a dependency in the package provides access to the Command Plugin and any targets to be linted add SwiftLint as a **_plugin_**, not a target dependency (which is what would be done if developing something against SwiftLint).

Reviewer 2: This is **_not_** for developing something against SwiftLint.

The existence of SwiftLint as a package dependency provides access to the Command Plugin and any targets to be linted add SwiftLint as a **_plugin_**, not a target dependency (which is what would be done if developing something against SwiftLint).

Reviewer: "Xcode Package Dependency" sounds like more than adding a plugin. That's what confused me. But that's actually the correct name I think. We should be very clear here, though, that SwiftLint must not be added as a target dependency. Again and again people report issues because of ignoring that. "It is not required" in the box below sounds to lax in that regard.

> Use the following link to add SwiftLint as a Package Dependnecy to an Xcode

Reviewer: Typo: Dependnecy

Author: Thank you!

project:

```
https://github.com/realm/SwiftLint
```

> [!IMPORTANT]
> It is not required to add the `SwiftLintFramework` library or the
> `swiftlint` executable to any targets, therefore, ensure `None`
> is selected when asked to choose package products.

### Homebrew

http://brew.sh

```
brew install swiftlint
```

### CocoaPods

https://cocoapods.org

Add the following to your `Podfile`:

```ruby
pod 'SwiftLint'
```

This will download the SwiftLint binaries and dependencies in `Pods/` during
your next `pod install` execution and will allow you to invoke it via
`${PODS_ROOT}/SwiftLint/swiftlint` in your Script Build Phases.

Installing via Cocoapods also enables pinning to a specific version of

> SwiftLint rather than simply the latest (which is the case with Homebrew).

Reviewer: Link to the Homebrew section?

Note that this will add the SwiftLint binaries, its dependencies' binaries, and
the Swift binary library distribution to the `Pods/` directory, so checking in

> this directory to SCM such as git is discouraged.

Reviewer: ```suggestion
this directory to SCM such as Git is discouraged.
```

By participating, you are expected to uphold this code. Please report
unacceptable behavior to [info@realm.io](mailto:info@realm.io).

> Switch Language:
> [中文](https://github.com/realm/SwiftLint/blob/main/README_CN.md)
> [한국어](https://github.com/realm/SwiftLint/blob/main/README_KR.md)

## Video Introduction

To get a high-level overview of SwiftLint, we encourage you to watch this
presentation recorded January 9th, 2017 by JP Simard (transcript provided):

[![Presentation](https://raw.githubusercontent.com/realm/SwiftLint/main/assets/presentation.svg)](https://youtu.be/9Z1nTMTejqU)

## Installation

### [Swift Package Manager](https://github.com/apple/swift-package-manager)

SwiftLint can be used as a [command plugin](#swift-package-command-plugin)
or a [build tool plugin](#swift-package-build-tool-plugins).

Add

```swift
.package(url: "https://github.com/realm/SwiftLint.git", from: "<version>")
```

to your `Package.swift` file to consume the latest release of SwiftLint
automatically or pin the dependency to a specific version:

```swift
.package(url: "https://github.com/realm/SwiftLint.git", exact: "<version>")
```

> > [!IMPORTANT]

Reviewer 2: I believe that can be a note instead. (allows the other important callouts to stand out more)
```suggestion
> [!NOTE]
```

bazel run -c opt @SwiftLint//:swiftlint
```

### Pre-Built Package

Download `SwiftLint.pkg` from the
[latest GitHub release](https://github.com/realm/SwiftLint/releases/latest) and
run it.

### From Source

> Cloning this project and run `make install`
> (Requires [bazel](https://bazel.build)).

Reviewer: ```suggestion
Make sure the build tool [Bazel](https://bazel.build) and a
recent [Swift toolchain](https://www.swift.org/download/) are
installed and all tools are discoverable in your `PATH`.

To build SwiftLint, clone this repository and run `make install`.
```

  other config file, e.g., `parent_config: path/to/.swiftlint.yml`.
- You can also consider the use of a
  [Run Script Build Phase](#xcode-run-script-build-phase) in place of the build
  tool plugin.

### Xcode Run Script Build Phase

If the build tool plugin does not work for your project setup or when
additional custom setup is required, SwiftLint can be added as a Run Script
Build Phase. This is useful when a project setup relies on the `--config`
SwiftLint option; or to lint all targets together in a single `swiftlint`
invocation. File inclusions and exclusions can be configured in the
[`.swiftlint.yml` configuration](#configuration).

To do this, add a custom script to a `Run Script` phase of the `Build Phases`
of the primary app target, after the `Compile Sources` phase. Use the
following script implementation:

```
if command -v swiftlint >/dev/null 2>&1
then
    swiftlint
else
    echo "warning: `swiftlint` command not found - See https://github.com/realm/SwiftLint#installation for installation instructions."
fi
```

<br>

> [!TIP]
> Uncheck `Based on dependency analysis` to run `swiftlint` on all incremental
> builds, suppressing the unspecified outputs warning.

> > [!IMPORTANT]

Reviewer 2: I believe that can be a note instead. (allows the other important callouts to stand out more)
```suggestion
> [!NOTE]
```

> https://github.com/realm/SwiftLint

That is because Homebrew on Apple Silicon installs the binaries into the
`/opt/homebrew/bin` folder by default. To instruct Xcode where to find
SwiftLint, you can either add `/opt/homebrew/bin` to the `PATH` environment
variable in your build phase:

```bash
if [[ "$(uname -m)" == arm64 ]]
then
    export PATH="/opt/homebrew/bin:$PATH"
fi

if command -v swiftlint >/dev/null 2>&1
then
    swiftlint
else
    echo "warning: `swiftlint` command not found - See https://github.com/realm/SwiftLint#installation for installation instructions."
fi
```

or you can create a symbolic link in `/usr/local/bin` pointing to the actual
binary:

```bash
ln -s /opt/homebrew/bin/swiftlint /usr/local/bin/swiftlint
```

**Additional Considerations**

If you wish to fix violations as well, your script could run
`swiftlint --fix && swiftlint` instead of just `swiftlint`. This will mean
that all correctable violations are fixed while ensuring warnings show up in
your project for remaining violations.

If you've installed SwiftLint via CocoaPods the script should look like this:

```bash
"${PODS_ROOT}/SwiftLint/swiftlint"
```

### Visual Studio Code

> To integrate SwiftLint with [vscode](https://code.visualstudio.com), install the

Reviewer: ```suggestion
To integrate SwiftLint with [Visual Studio Code](https://code.visualstudio.com), install the
```

bazel run -c opt @SwiftLint//:swiftlint
```

### Pre-Built Package

Download `SwiftLint.pkg` from the
[latest GitHub release](https://github.com/realm/SwiftLint/releases/latest) and
run it.

### From Source

Cloning this project and run `make install`
(Requires [bazel](https://bazel.build)).

## Setup

> [!IMPORTANT]
> While it may seem intuitive to run SwiftLint before compiling Swift source
> files to exit a build early when there are lint violations, it is important
> to understand that SwiftLint is designed to analyze valid source code that
> is compilable. Non-compiling code can very easily lead to unexpected and
> confusing results, especially when executing with `--fix`/`--autocorrect`
> command line arguments.

### Swift Package Command Plugin

> > [!IMPORTANT]

Reviewer 2: May this be "downgraded" to a note?
```suggestion
> [!NOTE]
```

bazel run -c opt @SwiftLint//:swiftlint
```

### Pre-Built Package

You can also install SwiftLint by downloading `SwiftLint.pkg` from the
[latest GitHub release](https://github.com/realm/SwiftLint/releases/latest) and
running it.

### From Source

> You can also build and install from source by cloning this project and running

Reviewer: Please start this sentence different from the previous one.

To maximize compatibility with the plugin, avoid project structures that require
the use of the `--config` option.

### Swift Package Projects

> > [!IMPORTANT]

Reviewer 2: May this be "downgraded" to a note?
```suggestion
> [!NOTE]
```

> Requires installing via [Swift Package Manager](#swift-package-manager).

Build tool plugins run when building each target. When a project has multiple
targets, the plugin must be added to the desired targets individually.

To do this, add the plugin to the target(s) to be linted as follows:

```swift
.target(
    ...
    plugins: [.plugin(name: "SwiftLintBuildToolPlugin", package: "SwiftLint")]
),
```

### Xcode Projects

> > [!IMPORTANT]

Reviewer 2: May this be "downgraded" to a note?
```suggestion
> [!NOTE]
```

> Requires installing via [Xcode Package Dependency](#xcode-package-dependency).

Build tool plugins run as a build phase of each target. When a project has
multiple targets, the plugin must be added to the desired targets individually.

To do this, add the `SwiftLintBuildToolPlugin` to the `Run Build Tool Plug-ins`
phase of the `Build Phases` for the target(s) to be linted.

> [!TIP]
> When using the plugin for the first time, be sure to trust and enable
> it when prompted. If a macros build warning exists, select it to trust
> and enable the macros as well.

For unattended use (e.g. on CI), package plugin and macro
validations can be disabled with either of the following:

* Using `xcodebuild` options:
  ```
  -skipPackagePluginValidation
  -skipMacroValidation
  ```
* Setting Xcode defaults:
  ```

>   defaults write com.apple.dt.Xcode IDESkipPackagePluginFingerprintValidatation -bool YES

Reviewer: Typo: `Validatation`

Author: Thank you!

- To use a config file located outside the package/project directory, a config 
  file may be added to that directory specifying a parent config path to the 
  other config file, e.g., `parent_config: path/to/.swiftlint.yml`.
- You can also consider the use of a
  [Run Script Build Phase](#xcode-run-script-build-phase) in place of the build
  tool plugin.

### Xcode Run Script Build Phase

If the build tool plugin does not work for your project setup or when
additional custom setup is required, SwiftLint can be added as a Run Script
Build Phase. This is useful when a project setup relies on the `--config`
SwiftLint option or to lint all targets together in a single `swiftlint`

> invocation.

Reviewer: ```suggestion
invocation. File inclusions and exclusions can be configured in the [`.swiftlint.yml` configuration](#configuration).
```

> [!IMPORTANT]

> > Compatibility is dependent upon the [installation](#installation) method that

Reviewer 2: Could be improved to something such as:

```
> [!IMPORTANT]
> Based upon the installation method used, the shell command syntax in the 
> Run Script Build Phase may be different or additional configuration could 
> be required. Refer to the [installation](#installation) instructions for 
> more information.
```

> is used.

Reviewer: ```suggestion
> is used.
```

To do this, add a custom script to a `Run Script` phase of the `Build Phases`

> of the primary app target, before the `Compile Sources` phase. Use the

Reviewer 2: Based on this language in another part of the readme, it sounds like there are pros/cons to linting before or after compiling. Since this is a general concern (not specific to any installation method for example), I wonder if this language could be improved and moved to a dedicated section between the Installation and Setup sections?

> You might want to move your SwiftLint phase directly before the 'Compile Sources' 
> step to detect errors quickly before compiling. However, SwiftLint is designed to 
> run on valid Swift code that cleanly completes the compiler's parsing stage. So 
> running SwiftLint before 'Compile Sources' might yield some incorrect results.

> of the primary app target, before the `Compile Sources` phase. Use the

Reviewer 2: What about this?
> While it may seem intuitive to run SwiftLint before compiling Swift source 
> files to exit a build early when there are lint violations, it is important 
> to understand that SwiftLint is designed to analyze valid source code that 
> is compilable. Non-compiling code can very easily lead to unexpected and 
> confusing  results, especially when executing with `--fix`/`--autocorrect` 
> command line arguments.

> of the primary app target, before the `Compile Sources` phase. Use the

Reviewer: Sounds good. Could be an IMPORTANT box in the Setup section.

- To use a config file located outside the package/project directory, a config 
  file may be added to that directory specifying a parent config path to the 
  other config file, e.g., `parent_config: path/to/.swiftlint.yml`.
- You can also consider the use of a
  [Run Script Build Phase](#xcode-run-script-build-phase) in place of the build
  tool plugin.

### Xcode Run Script Build Phase

If the build tool plugin does not work for your project setup or when
additional custom setup is required, SwiftLint can be added as a Run Script
Build Phase. This is useful when a project setup relies on the `--config`

> SwiftLint option or to lint all targets together in a single `swiftlint`

Reviewer 2: A semi-colon here might help with readability.
```suggestion
SwiftLint option; or to lint all targets together in a single `swiftlint`
```

invocation. File inclusions and exclusions can be configured in the
[`.swiftlint.yml` configuration](#configuration).

> > [!IMPORTANT]

Reviewer 2: Please place this important callout immediately following the `Uncheck Based on dependency analysis` tip.

> Based upon the installation method used, the shell command syntax in the 
> Run Script Build Phase may be different or additional configuration could 
> be required. Refer to the [installation](#installation) instructions for 
> more information.

To do this, add a custom script to a `Run Script` phase of the `Build Phases`
of the primary app target, after the `Compile Sources` phase. Use the
following script implementation:

```
if command -v swiftlint >/dev/null 2>&1
then
    swiftlint
else
    echo "warning: `swiftlint` command not found - See https://github.com/realm/SwiftLint#installation for installation instructions."
fi
```

<br>

> [!TIP]
> Uncheck `Based on dependency analysis` to run `swiftlint` on all incremental
> builds, suppressing the unspecified outputs warning.

> Xcode 15 made a significant change by setting the default value of the

Reviewer 2: ```suggestion
**Consideration for Xcode 15.0+**

Xcode 15 made a significant change by setting the default value of the
```

`ENABLE_USER_SCRIPT_SANDBOXING` build setting from `NO` to `YES`.
As a result, SwiftLint encounters an error related to missing file permissions,
which typically manifests as 
`error: Sandbox: swiftlint(19427) deny(1) file-read-data.`

To resolve this issue, it is necessary to manually set the
`ENABLE_USER_SCRIPT_SANDBOXING` setting to `NO` for the specific target that
SwiftLint is being configured for.

> If you installed SwiftLint via Homebrew on Apple Silicon, you might experience

Reviewer 2: ```suggestion
**Consideration for Apple Silicon**

If you installed SwiftLint via Homebrew on Apple Silicon, you might experience
```

this warning:

> warning: SwiftLint not installed, download from
> https://github.com/realm/SwiftLint

That is because Homebrew on Apple Silicon installs the binaries into the
`/opt/homebrew/bin` folder by default. To instruct Xcode where to find
SwiftLint, you can either add `/opt/homebrew/bin` to the `PATH` environment
variable in your build phase:

```bash
if [[ "$(uname -m)" == arm64 ]]
then
    export PATH="/opt/homebrew/bin:$PATH"
fi

if command -v swiftlint >/dev/null 2>&1
then
    swiftlint
else
    echo "warning: `swiftlint` command not found - See https://github.com/realm/SwiftLint#installation for installation instructions."
fi
```

or you can create a symbolic link in `/usr/local/bin` pointing to the actual
binary:

```bash
ln -s /opt/homebrew/bin/swiftlint /usr/local/bin/swiftlint
```

> 
> You might want to move your SwiftLint phase directly before the
> 'Compile Sources' step to detect errors quickly before compiling. However,
> SwiftLint is designed to run on valid Swift code that cleanly completes the
> compiler's parsing stage. So running SwiftLint before 'Compile Sources' might
> yield some incorrect results.

Reviewer 2: Remove since this is addressed at the beginning of the Setup section now.
```suggestion
```

> If you wish to fix violations as well, your script could run

Reviewer 2: ```suggestion
**Additional Considerations**

If you wish to fix violations as well, your script could run
```
